ecosystem/indexer/src/indexer/tailer.rs

```diff
@@ -0,0 +1,193 @@
+                        .is_ok()
+                    {
+                        fixed += 1;
+                    };
+                }
+                info!(
+                    "Fixed {}/{} previously errored versions for {}",
+                    fixed,
+                    err_count,
+                    processor2.name(),
+                );
+            });
+            tasks.push(task);
+        }
+        await_tasks(tasks).await;
+        info!("Fixing previously errored versions complete!");
+    }
+
+    /// Sets the version of the fetcher to the lowest version among all processors
+    pub async fn set_fetcher_to_lowest_processor_version(&self) -> u64 {
+        let mut lowest = u64::MAX;
+        for processor in &self.processors {
+            let max_version = processor.get_max_version().unwrap_or_default();
+            aptos_logger::debug!(
+                "Processor {} max version is {}",
+                processor.name(),
+                max_version
+            );
+            if max_version < lowest {
+                lowest = max_version;
+            }
+        }
+        aptos_logger::info!("Lowest version amongst all processors is {}", lowest);
+        self.set_fetcher_version(lowest).await;
+        lowest
+    }
+
+    pub async fn set_fetcher_version(&self, version: u64) -> u64 {
+        self.transaction_fetcher.lock().await.set_version(version);
+        aptos_logger::info!("Will start fetching from version {}", version);
+        version
+    }
+
+    pub async fn process_next(
+        &mut self,
+    ) -> anyhow::Result<Vec<Result<ProcessingResult, TransactionProcessingError>>> {
+        let txn = self.get_next_txn().await;
+        self.process_transaction(txn).await
+    }
+
+    pub async fn process_version(
+        &mut self,
+        version: u64,
+    ) -> anyhow::Result<Vec<Result<ProcessingResult, TransactionProcessingError>>> {
+        let txn = self.get_txn(version).await;
+        self.process_transaction(txn).await
+    }
+
+    pub async fn process_next_batch(
+        &mut self,
+        batch_size: u8,
+    ) -> Vec<anyhow::Result<Vec<Result<ProcessingResult, TransactionProcessingError>>>> {
+        let mut tasks = vec![];
+        for _ in 0..batch_size {
+            let mut self2 = self.clone();
+            let task = tokio::task::spawn(async move { self2.process_next().await });
+            tasks.push(task);
+        }
+        let results = await_tasks(tasks).await;
+        results
+    }
+
+    pub async fn process_transaction(
+        &self,
+        txn: Arc<Transaction>,
+    ) -> anyhow::Result<Vec<Result<ProcessingResult, TransactionProcessingError>>> {
+        let mut tasks = vec![];
+        for processor in &self.processors {
+            let txn2 = txn.clone();
+            let processor2 = processor.clone();
+            let task = tokio::task::spawn(async move {
+                processor2.process_transaction_with_status(txn2).await
+            });
```

Comment:
why not use join: https://rust-lang.github.io/async-book/06_multiple_futures/02_join.html

not sure if there is any perf implication but it feels kinda like a bit of boiler plate not to mention you're awaiting in a loop which from what I recall may result in serial execution

Reply:
I'll take your word for it ... async systems are magical